File: solutions/033_merge_k_sorted_list.go
package main

import "sort"

// ListNode for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

func mergeKLists(lists []*ListNode) *ListNode {
	array := []int{}
	// 读取所有链表的值至数组里
	for _, list := range lists {
		for list != nil {
			array = append(array, list.Val)
			list = list.Next
		}
	}
	sort.Ints(array) // 排序就完事了
	// array to list
	header := &ListNode{0, nil}
	tmp := header
	for _, v := range array {
		tmp.Next = &ListNode{v, nil}
		tmp = tmp.Next
	}
	return header.Next
}

File: solutions/__033_merge_k_sorted_list.java
import java.util.PriorityQueue;

public class __033_merge_k_sorted_list {
    public static void main(String[] args) {

    }

    public ListNode mergeKLists(ListNode[] lists) {
        PriorityQueue<ListNode> pq = new PriorityQueue<>((ListNode a, ListNode b) -> {
            return a.val - b.val;
        });
        ListNode Head = new ListNode();
        for (int i = 0; i < lists.length; i++) {
            if (lists[i] != null)
                pq.offer(lists[i]);
        }
        ListNode tmp = Head;
        while (!pq.isEmpty()) {
            ListNode node = pq.poll();
            if (node.next != null) {
                pq.offer(node.next);
            }
            tmp.next = new ListNode(node.val);
            tmp = tmp.next;
        }
        return Head.next;
    }
}

class ListNode {
    int val;
    ListNode next;

    ListNode() {
    }

    ListNode(int val) {
        this.val = val;
    }

    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }
}
